src/ui/handler/dblclick_zoom.js

```diff
@@ -71,15 +72,47 @@ class DoubleClickZoomHandler {
         if (!this.isEnabled()) return;
         if (e.points.length > 1) return;
 
+        const maxDelta = 30;
+
         if (!this._tapped) {
-            this._tapped = setTimeout(() => { this._tapped = null; }, 300);
+            this._tappedPoint = e.points[0];
+            this._tapped = setTimeout(() => { this._tapped = null; this._tappedPoint = null; }, 300);
         } else {
-            clearTimeout(this._tapped);
-            this._tapped = null;
-            this._zoom(e);
+            const newTap = e.points[0];
+            const firstTap = this._tappedPoint;
+
+            if (firstTap && Math.abs(firstTap.x - newTap.x) <= maxDelta && Math.abs(firstTap.y - newTap.y) <= maxDelta) {
+                e.originalEvent.preventDefault(); // prevent duplicate zoom on dblclick
+
+                const onTouchEnd = () => {
+                    if (this._tapped) { // make sure we are still within the timeout window
+                        this._zoom(e); // pass this touchstart event, as touchend events have no points
+                    }
+                    this._map.off('touchend', onTouchEnd);
+                    this._resetTapped();
+                };
+
+                const onTouchCancel = () => {
+                    this._map.off('touchend', onTouchEnd);
+                    this._map.off('touchcancel', onTouchCancel);
+                    this._resetTapped();
+                };
+
+                this._map.on('touchend', onTouchEnd);
+                this._map.on('touchcancel', onTouchCancel);
```

Comment:
To simplify this a bit, you can do `this._map.once(...)` instead of `on` — this way listener doesn't have to manually remove itself. It also ensures `touchcancel` listeners are not accumulated if they never happen.

Reply:
Ah, disregard the last sentence — it seems like we still need to make sure `touchcancel` listeners don't accumulate with every tap when touches are not cancelled.

Reply:
:+1: thanks for catching the missing `touchcancel` removal! I've now used `.once()` for both listeners and still manually remove the other listener in each handler, so I think all bases should be covered.